sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2557,24 +2557,45 @@ pi_result _pi_platform::populateDeviceCacheIfNeeded() {
       // Additionally we need to cache all sub-devices too, such that they
       // are readily visible to the piextDeviceCreateWithNativeHandle.
       //
+      std::vector<pi_device> PiSubDevices;
       pi_uint32 SubDevicesCount = 0;
       ZE_CALL(zeDeviceGetSubDevices,
               (Device->ZeDevice, &SubDevicesCount, nullptr));
 
-      auto ZeSubdevices = new ze_device_handle_t[SubDevicesCount];
-      ZE_CALL(zeDeviceGetSubDevices,
-              (Device->ZeDevice, &SubDevicesCount, ZeSubdevices));
+      if (SubDevicesCount) {
+        auto ZeSubdevices = new ze_device_handle_t[SubDevicesCount];
+        ZE_CALL(zeDeviceGetSubDevices,
+                (Device->ZeDevice, &SubDevicesCount, ZeSubdevices));
+
+        // Wrap the Level Zero sub-devices into PI sub-devices
+        // and add them to cache.
+        for (uint32_t I = 0; I < SubDevicesCount; ++I) {
+          std::unique_ptr<_pi_device> PiSubDevice(
+              new _pi_device(ZeSubdevices[I], this, Device.get()));
+          pi_result Result = PiSubDevice->initialize();
+          if (Result != PI_SUCCESS) {
+            delete[] ZeSubdevices;
+            return Result;
+          }
+          // save pointers to sub-devices for quick retrieval in the future.
+          Device->SubDevices.push_back(PiSubDevice.get());
+          PiDevicesCache.push_back(std::move(PiSubDevice));
+        }
+        delete[] ZeSubdevices;
+        PiSubDevices = Device->SubDevices;
+      } else {
+        // Level Zero will not report the sub-device on a single tile
+        // machine, but we still want that to be partitionable into
+        // multiple CCS-s. In this case the loop below will find out
+        // the CCS-s and report them as sub-devices of the root
+        // device.
+        SubDevicesCount = 1;
+        PiSubDevices.push_back(Device.get());
```

Comment:
This is the only new code in this PR, everything else is just restructuring to allow this new logic.